engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/Locking.java

```diff
@@ -0,0 +1,65 @@
+package org.enso.interpreter.instrument.execution;
+
+import java.io.File;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+
+/** Provides locking capabilities for the runtime server. */
+public interface Locking {
+
+  void assertWriteCompilationLock();
+
+  /**
+   * Executes `callable` while holding a compilation write lock
+   *
+   * @param where the class requesting the lock
+   * @param callable code to be executed while holding the lock
+   * @return the result of calling `callable` or null, if no result is expected
+   */
+  <T> T withWriteCompilationLock(Class<?> where, Callable<T> callable);
```

Comment:
This is what I wanted once, I think!

Reply:
https://github.com/enso-org/enso/pull/8270#discussion_r1393982296 yes.